lib/FilesHooks.php

```diff
@@ -1219,4 +1210,154 @@ protected function commitActivityTransaction(bool $shouldFlush): void {
+				$knownRules[$storageId] = [];
+			}
+
+			// caching rules based on storage+path to file
+			if (!array_key_exists($cachedMount['visiblePath'], $knownRules[$storageId])) {
+				$path = $cachedMount['path'];
+
+				// we need mountPoint and folderId to generate the correct path
+				try {
+					$node = $this->rootFolder->get($path);
+					$mountPoint = $node->getMountPoint();
+					$folderId = $mountPoint->getFolderId();
+					$path = substr($path, strlen($mountPoint->getMountPoint()));
+				} catch (\Exception $e) {
+					// in case of issue during the process, we can imagine the user have no access to the file
+					$usersToCheck[] = $cachedMount['userId'];
+					continue; // we'll catch rules on next user with access to the file
+				}
+
+				// we generate a list of path from top level of group folder to the file itself to get all rules
+				$paths = ['__groupfolders/' . $folderId];
+				while ($path !== '') {
+					$paths[] = '__groupfolders/' . $folderId . '/' . $path;
+					$path = dirname($path);
+					if ($path === '.' || $path === '/') {
+						$path = '';
+					}
+				}
+
+				// we might already know the rules for some path of the list
+				$paths = array_filter($paths, function (string $path) use ($knownRules, $storageId): bool {
+					if (array_key_exists($path, $knownRules[$storageId])) {
+						return false;
+					}
+
+					return true;
+				});
+
+				// we get and store the rules for each path from the list
+				$rulesPerPath = $ruleManager->getAllRulesForPaths($storageId, $paths);
+				foreach (array_keys($rulesPerPath) as $path) {
+					$rules = array_merge($rules, $rulesPerPath[$path]);
+				}
+
+				$knownRules[$storageId][$cachedMount['visiblePath']] = true;
+			}
+		}
+
+		// using each rules that disable read permission to generate a list of users
+		// that might not have access to fileId
+		foreach ($rules as $rule) {
+			if (($rule->getMask() & 1) === 0
+				|| ($rule->getPermissions() & 1) !== 0) {
+				continue; // not interested of rules with 'mask' not including read capability (1), or if 'permission' does
+			}
+
+			$mapping = $rule->getUserMapping();
+			$id = $mapping->getId();
+
+			// if mapping is about user
+			if ($mapping->getType() === 'user' && !in_array($id, $usersToCheck)) {
+				$usersToCheck[] = $id;
+			}
+
+			// if mapping is about group
+			if ($mapping->getType() === 'group') {
+				$group = $this->groupManager->get($id);
+				if ($group === null) {
+					continue;
+				}
+				$userIds = array_map(function (IUser $user): string {
+					return $user->getUID();
+				}, $group->getUsers());
+
+				// merge current user list with members of the group
+				$usersToCheck = array_values(array_unique(array_merge($usersToCheck, $userIds)));
+			}
+		}
+
+		// now that we have a list of 'unstable' users, we confirm they have no access to the file
+		$filteredUsers = [];
+		foreach ($usersToCheck as $userId) {
+			try {
+				if (count($this->rootFolder->getUserFolder($userId)->getById($fileId)) > 0) {
```

Comment:
Don't know how most groupfolder setups are and how often we end up in this code path, but if we now really set up 10k user storages when there is an negative rule, that is not going to scale and running out of memory pretty quickly.

Reply:
I don't remember how getById is setting up all mounts, it might be possible to use a shortcut and go straight to the group folder mount point name since we have it (assuming that it doesn't change in case of collisions!).
If the latter theory works, it might be possible to use `getUserfolder($userId)->get($pathToFile)` which might be more efficient.

I also wonder if we should tearDown the FS or if that's done automatically nowadays after some limit.

@icewind1991 can you advise?

Reply:
> I don't remember how getById is setting up all mounts, it might be possible to use a shortcut and go straight to the group folder mount point name since we have it (assuming that it doesn't change in case of collisions!).

The `getUserFolder` before will setUp the full user? Or at least load all the mount point informations (share table being contacted X times, etc)

Reply:
@nickvergessen yes and no. As far as I remember since NC 24 we have optimized the mount point setup and that part should be lazy. I hope @icewind1991 can confirm.

Reply:
I switched to `rootFolder->get(fullpath)`, but this might also limit the visibility on the access of the file as it limit the check on one path.

The fix was to merge result with the one from `$accessList['users']` that contains some other path to the targeted file.

please check last commit.

Reply:
@nickvergessen are you ok with the last changes ? (asking since you initiated the topic)

Reply:
I have no clue on the actual content. Just was scared of getUserFolder() inside a loop. Seems to be gone now, so not an issue anymore.